lib/Doctrine/ODM/MongoDB/DocumentManager.php

```diff
@@ -749,53 +749,43 @@ private function getDiscriminatorData(array $referenceMapping, ClassMetadata $cl
-                $pos = array_search($class->name, $referenceMapping['discriminatorMap']);
-                if ($pos !== false) {
-                    $discriminatorValue = $pos;
-                }
-            } else {
-                $discriminatorValue = $class->name;
+                $discriminatorMap = $referenceMapping['discriminatorMap'];
             }
         } else {
             $discriminatorField = $class->discriminatorField;
-            $discriminatorValue = isset($class->discriminatorValue) ? $class->discriminatorValue : $class->name;
+            $discriminatorValue = $class->discriminatorValue;
+            $discriminatorMap   = $class->discriminatorMap;
         }
 
-        if ($discriminatorField !== null) {
-            if ($discriminatorValue === null) {
-                @trigger_error(sprintf('Document class "%s" is unlisted in the discriminator map for reference "%s". This is deprecated and will throw an exception in 2.0.', $class->name, $referenceMapping['name']), E_USER_DEPRECATED);
-                $discriminatorValue = $class->name;
-            }
-
-            $discriminatorData = [$discriminatorField => $discriminatorValue];
-        } elseif (! isset($referenceMapping['targetDocument'])) {
-            $discriminatorField = $referenceMapping['discriminatorField'];
-
-            $discriminatorMap = null;
-            if (isset($referenceMapping['discriminatorMap'])) {
-                $discriminatorMap = $referenceMapping['discriminatorMap'];
-            }
+        if ($discriminatorField === null) {
+            return [];
+        }
 
-            if ($discriminatorMap === null) {
-                $discriminatorValue = $class->name;
-            } else {
-                $discriminatorValue = array_search($class->name, $discriminatorMap);
+        if ($discriminatorValue === null) {
+            if (! empty($discriminatorMap)) {
+                $discriminatorValue = null;
+                $pos = array_search($class->name, $discriminatorMap);
 
-                if ($discriminatorValue === false) {
-                    @trigger_error(sprintf('Document class "%s" is unlisted in the discriminator map for reference "%s". This is deprecated and will throw an exception in 2.0.', $class->name, $referenceMapping['name']), E_USER_DEPRECATED);
-                    $discriminatorValue = $class->name;
+                if ($pos !== false) {
+                    $discriminatorValue = $pos;
                 }
+            } else {
+                $discriminatorValue = $class->name;
             }
+        }
 
-            $discriminatorData = [$discriminatorField => $discriminatorValue];
+        if ($discriminatorValue === null) {
+            @trigger_error(sprintf('Document class "%s" is unlisted in the discriminator map for reference "%s". This is deprecated and will throw an exception in 2.0.', $class->name, $referenceMapping['name']), E_USER_DEPRECATED);
```

Comment:
What do you think of including the package name in deprecations?

```suggestion
            @trigger_error(sprintf('Document class "%s" is unlisted in the discriminator map for reference "%s". This is deprecated and will throw an exception in doctrine/mongodb-odm 2.0.', $class->name, $referenceMapping['name']), E_USER_DEPRECATED);
```